client/extensions/woocommerce/woocommerce-services/state/service-settings/reducer.js

```diff
@@ -13,7 +13,7 @@ import { WOOCOMMERCE_SERVICES_SERVICE_SETTINGS_UPDATE_FIELD } from '../action-ty
 const reducers = {};
 
 reducers[ WOOCOMMERCE_SERVICES_SERVICE_SETTINGS_UPDATE_FIELD ] = ( state, action ) => {
-	return objectPath.set( state, action.path, action.value );
+	return setWith( clone( state ), action.path, action.value, clone );
```

Comment:
clone is only a shallow clone. I'm guessing that since you're using clone with setWith, you're changing bits of the object past the first level, so you may need deepClone here instead.

ideally we'd check to see if the new value if different from the current value and return the current state if there's no change.

Reply:
The last argument of `setWith` is a function that will be called on every "branch" of the `state` object that's in `path`. So, essentially, these 2 ways are equivalent:
```js
return setWith( clone( state ), 'a.b.c', 42, clone );
```
```js
return {
    ...state,
    a: {
        ...state.a,
        b: {
            ...state.b,
            c: 42,
        },
    },
};
```

Maybe using `lodash/fp/set` would be less surprising than this? It also shallow clones the branches of the state that will be changed.

Reply:
we've historically avoides `lodash/fp` because it's a pretty heavy dependency (~14K gzipped).

another option would be https://github.com/kolodny/immutability-helper, which we use some other places in the app.

Reply:
and thanks for the explanation. I think my eyes glossed over at the first `clone` and didn't see the second one passed as the updater param.

Reply:
> we've historically avoides lodash/fp because it's a pretty heavy dependency (~14K gzipped).

Oh, I saw there were some usages in Calypso but turns out there are just in `example.jsx` files :)

> another option would be https://github.com/kolodny/immutability-helper, which we use some other places in the app.

Yeah i looked at it but it doesn't support passing an object path (`'a.b.c'`), so it wouldn'0t be as simple to use for this.

Reply:
ah, right, we'd have to convert from the string path to an object to use immutability-helper. 

looks good.

Reply:
> because it's a pretty heavy dependency

also because it's awkward :wink: